ldm/models/diffusion/ddim.py

```diff
@@ -14,17 +14,25 @@
 
 
 class DDIMSampler(object):
-    def __init__(self, model, schedule='linear', device='cuda', **kwargs):
+    def __init__(self, model, schedule='linear', device=None, **kwargs):
         super().__init__()
         self.model = model
         self.ddpm_num_timesteps = model.num_timesteps
         self.schedule = schedule
-        self.device = device
+        if not device:
+            if torch.cuda.is_available():
+                self.device = 'cuda'
+            elif torch.backends.mps.is_available():
+                self.device = 'mps'
+            else:
+                self.device = 'cpu'
+        else:
+            self.device = device
 
     def register_buffer(self, name, attr):
         if type(attr) == torch.Tensor:
             if attr.device != torch.device(self.device):
-                attr = attr.to(torch.device(self.device))
+                attr = attr.to(torch.float32).to(torch.device(self.device))
```

Comment:
A stacked `to` call like this could result in two copies of the underlying tensor. Instead pass the `dtype` and `device` arguments to one `to` call to limit the number of potential copies made like this:
```suggestion
                attr = attr.to(dtype=torch.float32, device=self.device)
```

Reply:
also, will this part affect CUDA and on-CPU users? might want a condition to isolate this change to MPS?

Reply:
If the user is in full precision, then all the tensors should already be float32. If the user is in half precision, then autocast should automatically be casting tensors to float16 and float32 as needed by the model layers.

If these tensors are currently float16 for GPU and CPU users using the default half precision, then this change would use more memory then the current implementation by forcing a cast to float32.